### ISSUE TITLE
Replace REST with GraphQL

### ISSUE THREAD

**Issue by Aurailus** [AuriServe/AuriServe#48]
I'll try to write up the formatting for this and update the issue when I do.

**Comment by Aurailus** [comment]
Schemas have been added for all existing endpoints. Resolvers need to be implemented for them now, this possibly needs to wait for #46.

**Comment by Aurailus** [comment]
After thinking about it, I'm not sure that GQL is actually a good fit for the administrator interface. The interface mainly works by querying chunks of data that form _complete_ object listings, e.g. the media page queries and receives _all_ of the media items with all their fields. GQL doesn't seem very well-suited for this task, and I'm wondering if potentially it isn't what should be used here. Thoughts, @octacian?

**Comment by octacian** [comment]
Right, GQL likes to force explicitness when fetching objects, requiring you to manually name every field even if you want all of them. That sounds bad, I know, but I don't think it is as bad as it sounds. Maybe you need to fetch all the media items with all of their fields now, but what happens when you add another field later on that is relevant only when serving that media item to the client? Now the admin has to fetch an additional field it doesn't need, or a second REST endpoint has to be created. Perhaps this isn't something that could ever be expected to happen with media, but there's a lot of other things the admin interfaces fetch besides media. I think it will make a lot more sense as we get into more complex data types, particularly when serving the public front-end.

GQL feels like a sort of 'you may not need it everywhere, especially at the start, but later on you'll really appreciate it' kind of library :smile:  And as such, I think we should stick with it. Reusable fragments can help ease the pain of fetching all the fields of large objects, or introspection can be used to dynamically build a query with all the fields of an object (can be done at compile-time too, no need for extra stuff on the client-side).

Remember as well, one of the major pros to GQL is its ability to fetch information from many different query endpoints in a single network request as opposed to making multiple requests, one for each REST endpoint. So when fully architectured, there should be benefit even in cases where the ability to pick and choose fields to fetch is unnecessary.
